glib/src/translate.rs

```diff
@@ -1081,56 +1166,104 @@ impl FromGlib<glib_sys::gboolean> for bool {
     }
 }
 
-impl FromGlib<u32> for char {
-    #[inline]
-    fn from_glib(val: u32) -> char {
-        char::from_u32(val).expect("Valid Unicode character expected")
-    }
-}
-
 impl FromGlib<i32> for Ordering {
     #[inline]
     fn from_glib(val: i32) -> Ordering {
         val.cmp(&0)
     }
 }
 
-impl FromGlib<u32> for Option<char> {
-    #[inline]
-    fn from_glib(val: u32) -> Option<char> {
-        match val {
-            0 => None,
-            _ => char::from_u32(val),
-        }
+/// Translate from a Glib type which can result in an undefined and/or invalid value.
+pub trait TryFromGlib<G: Copy>: Sized {
+    type Error;
+    fn try_from_glib(val: G) -> Result<Self, Self::Error>;
+}
+
+/// Error type for `TryFromGlib` when the Glib value is None.
+#[derive(Debug, PartialEq, Eq)]
+pub struct GlibNoneError;
+
+impl fmt::Display for GlibNoneError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "glib value is None")
     }
 }
 
-impl FromGlib<i32> for Option<u32> {
+impl std::error::Error for GlibNoneError {}
+
+impl<G: Copy, T: TryFromGlib<G, Error = GlibNoneError>> FromGlib<G> for Option<T> {
     #[inline]
-    fn from_glib(val: i32) -> Option<u32> {
-        if val >= 0 {
-            Some(val as u32)
-        } else {
-            None
+    fn from_glib(val: G) -> Option<T> {
+        T::try_from_glib(val).ok()
+    }
+}
+
+/// Error type for `TryFromGlib` when the Glib value can be None or invalid.
+#[derive(Debug, Eq, PartialEq)]
+pub enum GlibNoneOrInvalidError<I: Error> {
+    Invalid(I),
+    None,
+}
+
+impl<I: Error> GlibNoneOrInvalidError<I> {
+    /// Builds the `None` variant.
+    pub fn none() -> Self {
+        GlibNoneOrInvalidError::None
+    }
+
+    /// Returns `true` if `self` is the `None` variant.
+    // FIXME `matches!` was introduced in rustc 1.42.0, current MSRV is 1.40.0
+    // FIXME uncomment when CI can upgrade to 1.47.1
+    //#[allow(clippy::match_like_matches_macro)]
+    pub fn is_none(&self) -> bool {
+        match self {
+            GlibNoneOrInvalidError::None => true,
+            _ => false,
+        }
+    }
+
+    /// Returns `true` if `self` is the `Invalid` variant.
+    // FIXME `matches!` was introduced in rustc 1.42.0, current MSRV is 1.40.0
```

Comment:
No, we will increase it in the next gtk-rs release. Please make the update and don't worry about it. (In a following PR if you prefer though?)

Reply:
Will we go to 1.47 or 1.48 directly?

Note that if you start requiring 1.47 now then the GStreamer CI will fail forever until 1.47.1 or 1.48 is released (because of a clippy bug...).

Reply:
iirc, we'll need the 1.48 if we migrate to the clone! proc-macro.

Reply:
> iirc, we'll need the 1.48 if we migrate to the clone! proc-macro.

Fine with me once 1.48 is out and confirmed to work properly :)

Reply:
Ok, so I keep it like this for now and open an issue.

Reply:
Thanks!

Reply:
https://github.com/gtk-rs/gtk-rs/issues/17